### ISSUE TITLE
Upgrade ZF components. Zend_Service

### ISSUE THREAD

**Issue by okorshenko** [magento/magento2#9243]
**Description**
Upgrade components from ZF1 to ZF2

As a long term goal, we would like to eliminate knowledge about 3rd party libraries from Magento code base. Magento code still can use 3rd party libraries, but they must be wrapped by Magento interfaces and classes (adapters) so that 3rd party libraries can be easily substituted by newest versions or alternative implementations.

**Acceptance Criteria**

1. Magento interfaces are defined and can be used instead of ZF1 classes in Magento code
2. Default implementations for new interfaces are defined and configured as the preferences for new interfaces
3. Client code of the ZF1 components is refactored to use Magento interfaces
4. Zend classes are used as private services in Magento adapters
5. Backward compatibility is preserved
6. ZF1 components are updated to ZF2

**List of the components**

\# | Component Name <br/> (total usages count) | Module Name | Usages Count
------------- | ------------- | ------------- | -------------
1|Zend_Service (total: 1)|Framework\CurrencyInterface.php|1

**Comment by dverkade** [comment]
Hi @okorshenko 

I did some digging into this issue and here is what I found. 

As far as I can see the mention to \Zend_Service is incorrect and might be a bug/wrong in Zend Framwork. If you look into the \Zend_Currency class you see that it only allows for Zend_Currency_CurrencyInterface classes to be passed to the setService function. So the getService can only return a class which implements this interface.

```
public function setService($service)
    {
        if (is_string($service)) {
            #require_once 'Zend/Loader.php';
            if (!class_exists($service)) {
                $file = str_replace('_', DIRECTORY_SEPARATOR, $service) . '.php';
                if (Zend_Loader::isReadable($file)) {
                    Zend_Loader::loadClass($service);
                }
            }

            $service = new $service;
        }

        if (!($service instanceof Zend_Currency_CurrencyInterface)) {
            #require_once 'Zend/Currency/Exception.php';
            throw new Zend_Currency_Exception('A currency service must implement Zend_Currency_CurrencyInterface');
        }

        $this->_options['service'] = $service;
        return $this;
    }

```

I was also checking how Magento handles the fetching of the currency rates (that is what these functions are for) from and external service. I see this is done in the \Magento\Directory\Model\Currency\Import namespace and does not go truth this code.  

How do you want to handle this? 
- We can change the comment in the Framework\CurrencyInterface.php so that the comment will reflect the actual interface being returned, which is: Zend_Currency_CurrencyInterface 
- Maybe we can deprecate the getService and setService functions because they are not used. 
- Other solution?

**Comment by magento-engcom-team** [comment]
No activity. Closing the issue

**Comment by okorshenko** [comment]
Hi @dverkade 
Since this is `@api` interface we cannot remove any methods from the contract.
We need to modify doc blocks for the following methods:

```
    /**
     * Returns the set service class
     *
     * @return \Zend_Service
     */
    public function getService();

    /**
     * Sets a new exchange service
     *
     * @param string|\Magento\Framework\Locale\CurrencyInterface $service Service class
     * @return \Magento\Framework\CurrencyInterface
     */
    public function setService($service);
```

1. Remove `Magento\Framework\Locale\CurrencyInterface` from setService method and add align it with `\Zend_Currency` signatures
2. Remove `\Zend_Service` from getService method and add align it with `\Zend_Currency` signatures
3. Add `@deprecated` tags to both methods and add `@see` annotation to the methods. Please, reference corresponding interface from `\Magento\Directory\Model\Currency\Import` namespace that must be used instead these methods
4. We would like to cleanup entire interface, we can't deprecate other methods without suggested alternative implementations.  So let's stop on these two methods for now

It would be great if you can submit PR to `2.2-develop` branch.

Please, let me know if you have any questions.

**Comment by magento-team** [comment]
Hi @okorshenko. Thank you for your report. 
The issue has been fixed in magento/magento2#12957 by @dverkade in 2.3-develop branch
Related commit(s):
 - 68389d62965eb53c6533e51b375334ba691bdbb8

The fix will be available with the upcoming patch release.

**Comment by okorshenko** [comment]
Hi @okorshenko. Thank you for your report. 
The issue has been fixed in magento/magento2#12958 by @dverkade in 2.1-develop branch
Related commit(s):
 - abc7bc6b4f571a9131d97a732a8d0bc2a4271d5d

The fix will be available with the upcoming patch release.
